### ISSUE TITLE
Player Kicked on Server Change

### ISSUE THREAD

**Issue by MacTh3Mac** [CodeCrafter47/BungeeTabListPlus#727]
Environment:   Waterfall-1.20-537
Server: All Versions 1.12 through 1.20.1
Via-Version: 4.8.0
Via-Backwards: 4.8.0
Client Version:  Only affects 1.20.2 Vanilla Minecraft Client

BungeeTabListPlus: Release and Build 114 from CI Server.

Upon attempting to change servers within a bungee network the following error is generated:

<img width="1473" alt="Screenshot 2023-09-22 at 19 56 21" src="https://github.com/CodeCrafter47/BungeeTabListPlus/assets/34489434/729d8fd6-a504-4f1c-9c22-c5823b8971af">

`Internal Exception: io.nettyhandler.codec.DecoderException: z: Non [a-z0-9/._-] character in path of location: minecraft: BTLP4625264d 00`

This error occurs with or without the Bukkit Bridge installed.

This error is resolved by removing BTLP from Bungee.

No log entries made on either server or bungee.

Edit: added error in plain text to enable copy/paste or web search / crawling etc.

**Comment by cobrex1** [comment]
I also get the same error, using waterfall-537 & 538 also. If the client is on 1.20.1, no issues, but on 1.20.2 the error above occurs.

**Comment by gre3x** [comment]
Seems to be a more wide spread thing related to Bungee. I'm unsure if this would be fix on plugin side or Bungee side: https://github.com/SpigotMC/BungeeCord/issues/3519

**Comment by MacTh3Mac** [comment]
Appears to be unresolved by: 
https://github.com/SpigotMC/BungeeCord/commit/0509303fd37c2b236c8a3b5f5cea85cada937ad2

**Comment by MacTh3Mac** [comment]
Update when running https://github.com/SpigotMC/BungeeCord/commit/f9b75c4a3ad974d032437055ca90ef027d9b857f

Error Changes to:
`EncoderException: java.lang.IllegalArgumentException: Cannot get ID for packet class net.md_5.bungee.protocol.packet.Team in phase CONFIGURATION with direction TO_CLIENT @ io.netty.handler.codec.MessageToByteEncoder:125`

Error on Waterfall 538 remains as per previous comments when opening issue.
`Internal Exception: io.nettyhandler.codec.DecoderException: z: Non [a-z0-9/._-] character in path of location: minecraft: BTLP4625264d 00`

**Comment by MacTh3Mac** [comment]
Error Generated with BungeeCord:

https://pastebin.com/dUBUKpD8

**Comment by NoltoxGit** [comment]
I have the same problem.

**Comment by HippieBeak** [comment]
Looks like Bungee is suggesting that this might require a fix by BTLP rather than the Bungee side... https://github.com/SpigotMC/BungeeCord/issues/3519#issuecomment-1732435102

**Comment by gre3x** [comment]
@CodeCrafter47 looks like md5 is saying that this requires a fix in BTLP https://github.com/SpigotMC/BungeeCord/issues/3519#issuecomment-1732435102

**Comment by PrinceBunBun981** [comment]
I remember this being mentioned on the Discord and thought to give a small update here, Code is aware of the issues with 1.20.2 and will be looking into it, I don't have an exact ETA though.

**Comment by andrewkm** [comment]
Same problem happening here @CodeCrafter47 
https://i.imgur.com/kamuk6w.png

**Comment by Andre601** [comment]
Tagging him won't make things faster or change them. He does see the issues and gets notified. Tagging him only wastes your time.

**Comment by andrewkm** [comment]
> Tagging him won't make things faster or change them. He does see the issues and gets notified. Tagging him only wastes your time.

Typing out his name took me less than a second; my times not that valuable to consider a second wasted. Lol

**Comment by crumpkatrina** [comment]
We are also experiencing the same issue on our end.

**Comment by Phoenix616** [comment]
This is most likely because BLTP is either a) sending packets with the wrong IDs or b) sending them during the login or new configuration phase which of course doesn't support those. (Haven't further looked into this though)

**Comment by andfont** [comment]
please! something new? He's throwing me out to the players, thanks!

**Comment by dynamitestudios** [comment]
Anybody have a workaround until this is fixed? (Except telling players to use 1.20.1, removing BTLP, or downgrading the server.)

**Comment by HippieBeak** [comment]
I’m probably going to have to switch over to TAB if this isn’t resolved soon… in the BTLP discord they made it seem like it’s not even a guarantee that this will be fixed… just that CodeCrafter “may” look into this

**Comment by Andre601** [comment]
> I’m probably going to have to switch over to TAB if this isn’t resolved soon… in the BTLP discord they made it seem like it’s not even a guarantee that this will be fixed… just that CodeCrafter “may” look into this

Curse of open source.
This project is maintained by one guy in his free time, which is not a lot and usually is used by by more important stuff.
You can't and shouldn't expect updates being made fast nor that it is maintained forever. This isn't a project managed by employees of a company paid for this stuff...

**Comment by chrismwiggs** [comment]
> Curse of open source. This project is maintained by one guy in his free time, which is not a lot and usually is used by by more important stuff. You can't and shouldn't expect updates being made fast nor that it is maintained forever. This isn't a project managed by employees of a company paid for this stuff...

Came looking to see how likely it was that this issue would be looked at, to find this. If the project is abandoned, it would be nice to be marked as such. Not a demand or expectation of infinite support/service, just transparency is nice so when folks who depend on that work go to update, they know how to plan for the future.

Any chance of the dev or project lead commenting on this for clarity on even an intention to look at it? I get not having time to do it on demand but the communication is nice

**Comment by andrewkm** [comment]
> I’m probably going to have to switch over to TAB if this isn’t resolved soon… in the BTLP discord they made it seem like it’s not even a guarantee that this will be fixed… just that CodeCrafter “may” look into this

We’ve decided to just fork it privately and fix it ourselves along side a few other things we’ve wanted implemented for quite some time. May as well do so as well on your end. No use waiting tbh.

**Comment by Andre601** [comment]
> Came looking to see how likely it was that this issue would be looked at, to find this. If the project is abandoned, it would be nice to be marked as such. Not a demand or expectation of infinite support/service, just transparency is nice so when folks who depend on that work go to update, they know how to plan for the future.
> 
> Any chance of the dev or project lead commenting on this for clarity on even an intention to look at it? I get not having time to do it on demand but the communication is nice

It's not abandoned (And I in fact never said this in my message). It's just the simple fact that the main (And only) dev of the project does not have any time to spent on this project. This usually also includes not having time to write any comments. But I can assure you that CodeCrafter knows about this issues and wants to look into them. He just can't affort the time necessary for all of this.

It's something most people forget all the time when using plugins like BTLP. That a free resource may not have countless people behind it working on it.

> We’ve decided to just fork it privately and fix it ourselves along side a few other things we’ve wanted implemented for quite some time. May as well do so as well on your end. No use waiting if you need it quick. I'm sure eventually it'll be done here too.

Well, they may not have the knowledge? Just because you forked and fixed the problem doesn't mean other people have the knowledge to do so either...
Also, it would probably be a better move to also propose these fixes as a PR, as it could help fixing the issue faster, since CodeCrafter wouldn't need to spend time finding and fixing it himself and could instead just review the changes, maybe give some corrections and merge it if it is good.

**Comment by StanByes** [comment]
If is needed, I can try to check the issue and open a PR if I find a patch.

**Comment by Kas-tle** [comment]
Should be resolved with merge of https://github.com/CodeCrafter47/BungeeTabListPlus/pull/729 in latest update.

**Comment by MacTh3Mac** [comment]
Latest Update Resolves kick errors but introduces new issue.

Tab list formatting is lost on Server Change, Doesn't show tab list designated to server.  (issue on Bungeecord / waterfall latest)
 
Tested against:  
BungeeCord:  https://github.com/SpigotMC/BungeeCord/commit/1ef4d27dbea48a1d47501ad2be0d75e42cc2cc12
Waterfall: https://github.com/PaperMC/Waterfall/commit/ce48007a7d07ddc7d4d2dd842b70aa74e7034163
BTLP Version: 3.6.2 (Spigot Release) & BTLP 3.6.1-SNAPSHOT (https://ci.codecrafter47.de/job/BungeeTabListPlus/1141/)

**Comment by MacTh3Mac** [comment]
Additional Info Ref: https://github.com/CodeCrafter47/BungeeTabListPlus/issues/727#issuecomment-1742029133

Correct Tab List is displayed when 1st Joining,  Issue only occurs after Server Change.

**Comment by MacTh3Mac** [comment]
Additional info Ref: https://github.com/CodeCrafter47/BungeeTabListPlus/issues/727#issuecomment-1742029133

This display issue affects 1.20.2 Clients Only.     Lower Versions unaffected.

**Comment by andrewkm** [comment]
> Latest Update Resolves kick errors but introduces new issue.
> 
> Tab list formatting is lost on Server Change, Doesn't show tab list designated to server. (issue on Bungeecord / waterfall latest)
> 
> Tested against: BungeeCord: [SpigotMC/BungeeCord@1ef4d27](https://github.com/SpigotMC/BungeeCord/commit/1ef4d27dbea48a1d47501ad2be0d75e42cc2cc12) Waterfall: [PaperMC/Waterfall@ce48007](https://github.com/PaperMC/Waterfall/commit/ce48007a7d07ddc7d4d2dd842b70aa74e7034163) BTLP Version: 3.6.2 (Spigot Release) & BTLP 3.6.1-SNAPSHOT (https://ci.codecrafter47.de/job/BungeeTabListPlus/1141/)

Cannot reproduce.
Can you provide more information ,or an example.

**Comment by Phoenix616** [comment]
While the kick seems fixed to me it wont display any players after server switch.

![image](https://github.com/CodeCrafter47/BungeeTabListPlus/assets/5768781/8fa07564-88e2-480a-8cf0-65cfab0b6bf7)

I only use one global tablist though, not a per-server one.

**Comment by CodeCrafter47** [comment]
Does this issue involve ViaVersion? If so, can the issue be fixed by installing ViaVersion on the backend servers only, that is not on the proxy?

**Comment by Kas-tle** [comment]
We can resolve the error sent by @HippieBeak and @MacTh3Mac by checking if the connection is valid before queuing packets, but it seems that this error is not related to odd issues that occur after switching servers. Based on HippieBeak's note that it only occurs after leaving and rejoining, I suspect some data is somehow persisting in the packet queue after we terminate the connection and this leads to issues when a player rejoins until after proxy restart.

**Comment by MacTh3Mac** [comment]
> Does this issue involve ViaVersion? If so, can the issue be fixed by installing ViaVersion on the backend servers only, that is not on the proxy?

We have never used Via-Version on proxies,  far too many bugs.

**Comment by MacTh3Mac** [comment]
> but it seems that this error is not related to odd issues that occur after switching servers. Based on HippieBeak's note that it only occurs after leaving and rejoining, I suspect some data is somehow persisting in the packet queue after we terminate the connection and this leads to issues when a player rejoins until after proxy restart.

The tab list display issues can be recreated on a fresh proxy, busy proxy, empty proxy. on 1st join (proxy) and rejoin (proxy without restart).  It appears to be inherent in all situations using a 1.20.2 client.  And only on subsequent servers. Primary connection irrespective of which backend server the player arrives on always displays the correct tab list.  It's only after a backend server charge that all future tablists remain mis-displayed, including if you return back to the original connected backend server.  

Feel free to hop around our network on a 1.20.2 client and the issue should become immediately apparent. Details in Screenshot in prior replies.

**Comment by HippieBeak** [comment]
I cannot reproduce it on first join.

**Comment by MacTh3Mac** [comment]
Resolved By:  https://github.com/CodeCrafter47/BungeeTabListPlus/commit/a05beada9f8bf0904085637ccc0aa2ec3ddc8fd7

New Issues on incorrect display moved to new issue #731
